### DISCUSSION TITLE
OverrideElement Color Exploded View

### DISCUSSION THREAD

**Discussion by Paolo-lab** [iTwin/itwinjs-core#5564] [Q&A] [answered]
Hello everyone good morning 
I'm struggling with emphasizing some elements in my view
through the overrideElements of the EmphasizeElements class
I can color my elements etc. etc. and so far everything is ok.
At this point I would like the emphasized elements to Exploded like in the example https://www.itwinjs.org/sandboxes/iTwinPlatform/Exploded%20View
when I do the expansion the previously highlighted colors disappear I also implemented the emphasis again for this class but it doesn't seem to work Do you have any ideas on how to help me

Thanks for the precious help

**Comment by pmconne** [comment]
To understand the behavior you're observing, see bullet point 4 of the sample's "Overview" section, and line 335+ of ExplodeTile.ts.

> 4. Now that the modified elements are being drawn, the old, static elements need to be hidden. This is done with two parts working together. First, a [FeatureOverrideProvider](https://www.itwinjs.org/reference/core-frontend/views/featureoverrideprovider/) added to the view monitors the modified elements. Once they have tile content loaded, the ElementIds are marked to never be drawn using the [FeatureOverride.setNeverDrawnSet](https://www.itwinjs.org/reference/core-common/rendering/featureoverrides/) method. The second part is a [FeatureAppearanceProvider](https://www.itwinjs.org/reference/core-common/rendering/featureappearanceprovider/) being applied in the previously mentioned call to [RenderContext](https://www.itwinjs.org/reference/core-frontend/rendering/rendercontext/).createGraphicBranch. This provider ensures the elements relative to our tile tree are drawn by return a default [FeatureAppearance](https://www.itwinjs.org/reference/core-common/rendering/featureappearance/) for our modified elements.

You will need to modify your own code to apply your color overrides to the temporary graphics produced by ExplodeTile.

**Comment by Paolo-lab** [comment]
Hello everyone good morning I'm struggling with the extended of the elements, I had difficulty in maintaining the color of the elements once expanded, but thanks to you I was able to understand how to solve the problem, now I have another unexpected problem.
I can select the elements to expand via a <table>, but once expanded and then removed from the expansion list, the element disappears from my view.

the logic of the expansion from what i understand is to hide the element to expand and create a new one to edit, but how do i go back to the original element

thanks for the precious help

**Reply by pmconne** [reply, marked as answer]
From the previously provided snippet:

> First, a [FeatureOverrideProvider](https://www.itwinjs.org/reference/core-frontend/views/featureoverrideprovider/) added to the view monitors the modified elements. Once they have tile content loaded, the ***ElementIds are marked to never be drawn using the [FeatureOverride.setNeverDrawnSet](https://www.itwinjs.org/reference/core-common/rendering/featureoverrides/) method***.

If you want the original element to draw, you must remove it from the "never drawn" set.

**Comment by Paolo-lab** [comment]
Hello everyone
I'm always grappling with the expansion of the elements, thanks to your help I solved many of my problems (element colors, original elements)

I think I understand that in **explodeTile.ts** the "**compareTileTreeIds**" method compares the tile old and the tile new
comparing the number of elements
what I don't understand is when "**createTileTree**" comes into play
to reload data? when old Tile and new Tile are different ?

it seems like a random thing but I don't think it's like that, but it remains complex for me to understand how it works.

Thanks for the precious help

**Reply by pmconne** [reply]
> what I don't understand is when "createTileTree" comes into play

To instantiate the `TileTree` with the specified Id.
You register a [TiledGraphicsProvider](https://www.itwinjs.org/learning/display/tiledgraphicsprovider/) with the viewport. It provides access to one (in this example) [TileTreeReference](https://www.itwinjs.org/reference/core-frontend/tiles/tiletreereference/). When [creating the scene](https://www.itwinjs.org/learning/display/tiles/#scene-creation) we ask your tile tree reference to draw itself, it obtains the [TileTree](https://www.itwinjs.org/reference/core-frontend/tiles/tiletree/) it requires to do so. If that tile tree hasn't already been created (it won't have been, the first time we draw your tile tree reference), the tile tree reference will ask its [TileTreeSupplier](https://www.itwinjs.org/reference/core-frontend/tiles/tiletreesupplier/) to instantiate it via `createTileTree`. We will then cache that tree by its Id for subsequent reuse.

At some point in the future, we may discard the cached tile tree, because we haven't been asked to draw it for a while - maybe because you turned off your explode decorator. If you later turn the decorator back on, we will once again invoke `createTileTree`.
